### ISSUE TITLE
How to capture 401/403 push replication errors

### ISSUE THREAD

**Issue by vicpinm** [couchbase/couchbase-lite-java-core#1664]
Hi there! Sometimes I get differents pull replication errors, like Sync: {error=forbidden, id=027a5148-73e5-4893-9d29-80f1b216489b, reason=missing channel access, status=403}: _bulk_docs got an error: com.couchbase.lite.replicator.PusherInternal. I need to capture theese events to retry the save after x seconds. Pull replicator has addChangeListener method, but I don't receive these error in that listener. I'm only able to see these errors in the logcat. Is there any way to capture these errors?

**Comment by hideki** [comment]
Hi @vbenso,

`403` is a permanent error.  The error should be notified through `ChangeListener`. Can you check the return value from `ChangeEvent.getError()` method?

https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/replicator/Replication.java#L676

**Comment by vicpinm** [comment]
Hi @hideki, I have log every changeevent that arrives to changeEventListener, but I dont receive any 403 nor 401 change event. For example, when my session is expired, I can see in the logcat a lot of 401 errors:

ChangeTracker{http://couch.hablemos.com:4999/hablemos_test_dev_02, OneShot, @18eba06}: Change tracker got error 401
ChangeTracker: ChangeTracker{http://couch.hablemos.com:4999/hablemos_test_dev_02, OneShot, @d1738c7}: Change tracker got error 401
ChangeTracker: ChangeTracker{http://couch.hablemos.com:4999/hablemos_test_dev_02, OneShot, @783fb08}: Change tracker got error 401
ChangeTracker: ChangeTracker{http://couch.hablemos.com:4999/hablemos_test_dev_02, OneShot, @9de82a1}: Change tracker got error 401

But I don't receive any changeEvent with that error. I do receive other change events with different replication status (IDLE, STOP..). 

I'm using 1.4.0 version.

**Comment by hideki** [comment]
Hi @vicpinm,

We need to reproduce the problem. To do it, we need some information.
- What server do you use? Sync Gateway or CouchDB?  What version?
- Scenario how to reproduce the problem. Middle of oneshot (non-continuous) pull replication, logged in account lose the access to `channel`?

Thanks!

**Comment by vicpinm** [comment]
Hi @hideki 

We use Sync Gateway version 1.3. Is there any difference if we use CouchDB? I don't have clear the difference.

The problem appears with continuous replication, in a logged account which tries to replicante an object throw a channel in which I don't have permissions, so I get a 403 error. That is right, the problem is that I need to be notified of that event. I can see a locat line with that error:

Sync: {error=forbidden, id=027a5148-73e5-4893-9d29-80f1b216489b, reason=missing channel access, status=403}: _bulk_docs got an error: com.couchbase.lite.replicator.PusherInternal.

But I am not notified with a change event in my change listener. The same problem appears when my session is expired. I can see several 401 errors in logcat, but my listener is not invoked with that error.

**Comment by sebastianmacarescu** [comment]
Any news on this?

**Comment by sebastianmacarescu** [comment]
Hi @hideki 
I have managed to solve the issue with 401 errors by cloning the project and adding
`
if (tracker.getLastError() != null) {
                        setError(tracker.getLastError());
                    }
                    else {
                        setError(new Throwable(msg));
                    }
`
before this line: https://github.com/couchbase/couchbase-lite-java-core/blob/e276993f4fdd4beced1aea942b6a266eb848c648/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L939

**Comment by vicpinm** [comment]
Any updates with this?

**Comment by hideki** [comment]
Notes: 

CBL iOS implementation: Pull replicator obtains ChangeTracker's error variable from ChangeTracker when ChangeTracker stopped. But CBL Android implementation does not.

https://github.com/couchbase/couchbase-lite-ios/blob/86c63659b56d87406eac3b1e4db631fe67098139/Source/CBLRestPuller.m#L270-L281
```objective-c
// The change tracker reached EOF or an error.
- (void) changeTrackerStopped:(CBLChangeTracker *)tracker {
    if (tracker != _changeTracker)
        return;
    NSError* error = tracker.error;
    LogTo(Sync, @"%@: ChangeTracker stopped; error=%@", self, error.my_compactDescription);
    
    BOOL continous = _changeTracker.continuous;
    _changeTracker = nil;
    
    if (error) {
        if (CBLIsOfflineError(error))
            [self goOffline];
        else if (!self.error)
            self.error = error;
    }
    
    [_batcher flushAll];
    if (!continous)
        [self asyncTasksFinished: 1]; // balances -asyncTaskStarted in -startChangeTracker
    if (!_caughtUp)
        [self asyncTasksFinished: 1]; // balances -asyncTaskStarted in -beginReplicating
}
```

https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L922-L967

Solution: In case ChangeTracker is stopped with an error, pull replicator obtains error from ChangeTracker, and handles it accordingly.

**Comment by hideki** [comment]
Ref: CBSE-4462

**Comment by hideki** [comment]
Hi @vicpinm and @sebastianmacarescu,

We build the developer build with the fix for this issue. The solution is almost same with @sebastianmacarescu's suggestion. If you have time, could you please try the developer build?
The version number is 1.4.2-80. Please refer the following link for installation.

Instruction:
https://github.com/couchbase/couchbase-lite-android#adding-couchbase-lite-to-your-gradle-project

Couchbase custom maven repo:
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.4.2-80/

**Comment by vicpinm** [comment]
Hi @hideki, I just realized that I made a mistake when I wrote the issue. The problem is in the PUSH replicator, not pull replicator, as I happens when I save a new document, and I get a 403 error after that. I thought that error was notified through the pull replicator, but I think I was wrong. Sorry about that :(

**Comment by hideki** [comment]
NOTES: How to reproduce this error.

> 1. Create ACL doc (which gives access to user to certain channel)
> 2. Write doc to that channel
>
> In this scenario, if 2 is processed by SG before 1, then a 403 error is thrown because ACL access rule has not yet been created.

**Comment by hideki** [comment]
@vicpinm,

The fix, to remove the lines to ignore 401 or 403 error for push replication, causes the behavior change. It could break the applications that rely on current behavior. So we can not fix in short period of time. We will fix this for 1.5.0 release.
